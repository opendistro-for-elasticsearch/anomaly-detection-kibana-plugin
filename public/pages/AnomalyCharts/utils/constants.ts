/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

export const INITIAL_ANOMALY_SUMMARY = {
    anomalyOccurrence: 0,
    minAnomalyGrade: 0.0,
    maxAnomalyGrade: 0.0,
    minConfidence: 0.0,
    maxConfidence: 0.0,
    lastAnomalyOccurrence: '',
};


export enum ANOMALY_CHART_TITLE {
    SAMPLE_ANOMALY_HISTORY = 'Sample anomaly history',
    ANOMALY_HISTORY = 'Sample anomaly history',
}

export enum CHART_FIELDS {
    PLOT_TIME = 'plotTime',
    ANOMALY_GRADE = 'anomalyGrade',
    CONFIDENCE = 'confidence',
    DATA = 'data',
}

export enum CHART_COLORS {
    ANOMALY_GRADE_COLOR = '#D13212',
    FEATURE_DATA_COLOR = '#16191F',
    CONFIDENCE_COLOR = '#017F75',
}

export const FEATURE_CHART_THEME = {
    lineSeriesStyle: {
        line: {
            strokeWidth: 2,
            visible: true,
            opacity: 0.5,
        },
        point: {
            visible: true,
            stroke: CHART_COLORS.FEATURE_DATA_COLOR,
        },
    },
};

export const DATE_PICKER_QUICK_OPTIONS = [
    { start: 'now-24h', end: 'now', label: 'last 24 hours' },
    { start: 'now-7d', end: 'now', label: 'last 7 days' },
    { start: 'now-30d', end: 'now', label: 'last 30 days' },
    { start: 'now-90d', end: 'now', label: 'last 90 days' },

    { start: 'now/d', end: 'now', label: 'Today' },
    { start: 'now/w', end: 'now', label: 'Week to date' },
    { start: 'now/M', end: 'now', label: 'Month to date' },
    { start: 'now/y', end: 'now', label: 'Year to date' },
];

export enum LIVE_CHART_CONFIG {
    REFRESH_INTERVAL_IN_SECONDS = 30 * 1000, //poll anomaly result every 30 seconds
    MONITORING_INTERVALS = 60,
}
