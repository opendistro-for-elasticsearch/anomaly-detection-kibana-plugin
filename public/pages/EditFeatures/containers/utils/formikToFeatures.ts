/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { snakeCase, cloneDeep } from 'lodash';
import {
  Detector,
  FeatureAttributes,
  FEATURE_TYPE,
  UiFeature,
} from '../../../../models/interfaces';

import { AggregationOption } from '../../models/types';

export interface FeaturesFormikValues {
  featureId: string;
  featureName: string;
  featureType: FEATURE_TYPE;
  featureEnabled: boolean;
  aggregationQuery: string;
  aggregationBy?: string;
  aggregationOf?: AggregationOption[];
  newFeature?: boolean;
}

export function prepareDetector(
  values: FeaturesFormikValues[],
  ad: Detector,
  forPreview: boolean = false
): Detector {
  const detector = cloneDeep(ad);
  const featureAttributes = formikToFeatures(values, forPreview);

  return {
    ...detector,
    featureAttributes: [...featureAttributes],
    uiMetadata: {
      ...detector.uiMetadata,
      features: { ...formikToUIMetadata(values) },
    },
  };
}

export function formikToSimpleAggregation(value: FeaturesFormikValues) {
  if (
    value.aggregationBy &&
    value.aggregationOf &&
    value.aggregationOf.length > 0
  ) {
    return {
      [value.featureName]: {
        [value.aggregationBy]: { field: value.aggregationOf[0].label },
      },
    };
  } else {
    return {};
  }
}

export function formikToAggregation(values: FeaturesFormikValues) {
  if (values.featureType === FEATURE_TYPE.SIMPLE) {
    return values.aggregationBy &&
      values.aggregationOf &&
      values.aggregationOf.length > 0
      ? {
          [snakeCase(values.featureName)]: {
            [values.aggregationBy]: { field: values.aggregationOf[0].label },
          },
        }
      : {};
  }
  return JSON.parse(values.aggregationQuery);
}

export function formikToFeatures(
  values: FeaturesFormikValues[],
  forPreview: boolean
) {
  const featureAttribute = formikToFeatureAttributes(values, forPreview);
  return featureAttribute;
}

export function formikToUIMetadata(values: FeaturesFormikValues[]) {
  // TODO:: Delete Stale metadata if name is changed
  let features: {
    [key: string]: UiFeature;
  } = {};
  values.forEach(value => {
    if (value.featureType === FEATURE_TYPE.SIMPLE) {
      features[value.featureName] = {
        featureType: value.featureType,
        aggregationBy: value.aggregationBy,
        aggregationOf:
          value.aggregationOf && value.aggregationOf.length
            ? value.aggregationOf[0].label
            : undefined,
      };
    } else {
      features[value.featureName] = {
        featureType: value.featureType,
      };
    }
  });
  return features;
}

function formikToFeatureAttributes(
  values: FeaturesFormikValues[],
  forPreview: boolean
): FeatureAttributes[] {
  return values.map(function(value) {
    const id = forPreview
      ? value.featureId
      : value.newFeature
      ? undefined
      : value.featureId;
    return {
      featureId: id,
      featureName: value.featureName,
      featureEnabled: value.featureEnabled,
      importance: 1,
      aggregationQuery: formikToAggregation(value),
    };
  });
}
