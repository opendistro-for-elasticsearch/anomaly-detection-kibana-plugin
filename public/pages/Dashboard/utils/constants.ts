/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import {
  AD_DOC_FIELDS,
  SORT_DIRECTION,
} from '../../../../server/utils/constants';
import { ANOMALY_RESULT_INDEX, MAX_DETECTORS } from '../../../utils/constants';

export const TIME_RANGE_OPTIONS = [
  { value: '24h', text: 'Last 24 hours' },
  { value: '7d', text: 'Last 7 days' },
];

export const GET_RECENT_ANOMALOUS_DETECTORS_QUERY = {
  index: ANOMALY_RESULT_INDEX,
  size: 10,
  query: {
    bool: {
      must: [
        {
          range: {
            [AD_DOC_FIELDS.ANOMALY_GRADE]: {
              gt: 0.0,
            },
          },
        },
      ],
      must_not: [
        {
          exists: {
            field: AD_DOC_FIELDS.ERROR,
          },
        },
      ],
    },
  },
  sort: {
    [AD_DOC_FIELDS.DATA_START_TIME]: SORT_DIRECTION.DESC,
  },
  collapse: {
    field: AD_DOC_FIELDS.DETECTOR_ID,
  },
};

export const GET_RECENT_ANOMALY_RESULT_QUERY = {
  range: {
    [AD_DOC_FIELDS.DATA_START_TIME]: {
      gte: 'now-30m',
    },
  },
  size: 30,
  sortField: AD_DOC_FIELDS.DATA_START_TIME,
  from: 0,
  sortDirection: SORT_DIRECTION.DESC,
};

export const TIME_NOW_LINE_STYLE = {
  line: {
    strokeWidth: 1,
    stroke: '#3F3F3F',
    dash: [1, 2],
    opacity: 0.8,
  },
};

export const GET_ALL_DETECTORS_QUERY_PARAMS = {
  from: 0,
  search: '',
  indices: '',
  size: MAX_DETECTORS,
  sortDirection: SORT_DIRECTION.ASC,
  sortField: 'name',
};

export const ALL_DETECTORS_MESSAGE = 'All detectors';
export const ALL_DETECTOR_STATES_MESSAGE = 'All detector states';
export const ALL_INDICES_MESSAGE = 'All indices';
