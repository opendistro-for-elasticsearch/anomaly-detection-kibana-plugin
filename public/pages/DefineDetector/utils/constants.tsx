/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { FILTER_TYPES, UIFilter } from '../../../models/interfaces';
import { OPERATORS_MAP } from '../../DefineDetector/components/DataFilterList/utils/constant';
import { DetectorDefinitionFormikValues } from '../../DefineDetector/models/interfaces';

export const EMPTY_UI_FILTER: UIFilter = {
  filterType: FILTER_TYPES.SIMPLE,
  fieldInfo: [],
  operator: OPERATORS_MAP.IS,
  fieldValue: '',
  query: '',
  label: '',
};

export const INITIAL_DETECTOR_DEFINITION_VALUES: DetectorDefinitionFormikValues = {
  name: '',
  description: '',
  index: [],
  filters: [],
  filterQuery: JSON.stringify({ bool: { filter: [] } }, null, 4),
  timeField: '',
  interval: 10,
  windowDelay: 1,
};
