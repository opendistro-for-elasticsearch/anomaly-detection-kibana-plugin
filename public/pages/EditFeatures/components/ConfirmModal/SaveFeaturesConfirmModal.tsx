/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import {
  EuiFlexItem,
  EuiFlexGroup,
  EuiText,
  EuiRadioGroup,
} from '@elastic/eui';
import React from 'react';
// @ts-ignore
import { toastNotifications } from 'ui/notify';
//@ts-ignore
import chrome from 'ui/chrome';
import { ConfirmModal } from '../../../DetectorDetail/components/ConfirmModal/ConfirmModal';
import { SAVE_FEATURE_OPTIONS } from '../../utils/constants';

interface SaveFeaturesConfirmModalProps {
  readyToStartAdJob: boolean;
  saveFeatureOption: SAVE_FEATURE_OPTIONS;
  onClose(): void;
  onCancel(): void;
  onConfirm(): void;
  onOptionChange(id: string): void;
}

export function SaveFeaturesConfirmModal(props: SaveFeaturesConfirmModalProps) {
  const startAdJobOptions = (disableStartAdJob: boolean) => {
    return [
      {
        id: SAVE_FEATURE_OPTIONS.START_AD_JOB,
        label: 'Automatically run detector (Recommended)',
        disabled: disableStartAdJob,
      },
      {
        id: SAVE_FEATURE_OPTIONS.KEEP_AD_JOB_STOPPED,
        label: 'Manually start the detector at a later time',
      },
    ];
  };

  const confirmModalDescription = () => (
    <EuiFlexGroup direction="column">
      <EuiFlexItem grow={false}>
        <EuiText>
          <p>
            The detector is currently stopped. To receive accurate and real-time
            anomalies, the detector need to start and collect sufficient data to
            include your latest change. The earlier the detector starts running,
            the sooner the anomalies will be available.
          </p>
        </EuiText>
      </EuiFlexItem>
      <EuiFlexItem grow={false}>
        <EuiRadioGroup
          name="start ad radio group"
          options={startAdJobOptions(!props.readyToStartAdJob)}
          idSelected={props.saveFeatureOption}
          onChange={props.onOptionChange}
        />
      </EuiFlexItem>
    </EuiFlexGroup>
  );

  return (
    <ConfirmModal
      title="Automatically start the detector?"
      description={confirmModalDescription()}
      confirmButtonText="Confirm"
      confirmButtonColor="primary"
      onClose={props.onClose}
      onCancel={props.onCancel}
      onConfirm={props.onConfirm}
    />
  );
}
