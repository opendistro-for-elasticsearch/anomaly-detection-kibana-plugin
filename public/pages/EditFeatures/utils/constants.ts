/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { FEATURE_TYPE } from '../../../models/interfaces';
import { FeaturesFormikValues } from '../containers/utils/formikToFeatures';

export const FEATURE_TYPES = [
  { text: 'Custom Aggregation', value: FEATURE_TYPE.CUSTOM },
  { text: 'Defined Aggregation', value: FEATURE_TYPE.SIMPLE },
];

export const FEATURE_TYPE_OPTIONS = [
  { text: 'Field value', value: FEATURE_TYPE.SIMPLE },
  { text: 'Custom expression', value: FEATURE_TYPE.CUSTOM },
];

export enum SAVE_FEATURE_OPTIONS {
  START_AD_JOB = 'start_ad_job',
  KEEP_AD_JOB_STOPPED = 'keep_ad_job_stopped',
}

export const AGGREGATION_TYPES = [
  { value: 'avg', text: 'average()' },
  { value: 'sum', text: 'sum()' },
  { value: 'min', text: 'min()' },
  { value: 'max', text: 'max()' },
];

export const FEATURE_FIELDS = [
  'featureName',
  'aggregationOf',
  'aggregationBy',
  'aggregationQuery',
];

export const INITIAL_VALUES: FeaturesFormikValues = {
  featureId: '',
  featureName: '',
  featureEnabled: true,
  featureType: FEATURE_TYPE.SIMPLE,
  aggregationQuery: JSON.stringify(
    { aggregation_name: { sum: { field: 'field_name' } } },
    null,
    4
  ),
  aggregationBy: '',
  aggregationOf: [],
};
