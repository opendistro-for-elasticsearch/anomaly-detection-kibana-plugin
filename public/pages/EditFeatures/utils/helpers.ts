/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { DATA_TYPES } from '../../../utils/constants';
import { FEATURE_TYPE, FeatureAttributes, Detector } from '../../../models/interfaces';
import { v4 as uuidv4 } from 'uuid';
import { get } from 'lodash';
import { FeaturesFormikValues } from '../containers/utils/formikToFeatures';

export const getNumberFields = (allFields: { [key: string]: string[] }) =>
  [DATA_TYPES.NUMBER]
    .map(dataType =>
      allFields[dataType]
        ? {
            label: dataType,
            options: allFields[dataType].map(field => ({
              label: field,
              type: dataType,
            })),
          }
        : []
    )
    .filter(Boolean);

export const initialize_feature = () => ({
  featureId: uuidv4(),
  featureName: undefined,
  featureType: FEATURE_TYPE.SIMPLE,
  featureEnabled: true,
  importance: 1,
  aggregationQuery: JSON.stringify(
    {
      aggregation_name: { sum: { field: 'field_name' } },
    },
    null,
    4
  ),
  newFeature: true,
});

export const validateFeatures = (values: any) => {
  const featureList = get(values, 'featureList', []);
  let featureNameCount = new Map<string, number>();

  featureList.forEach((attribute: FeatureAttributes) => {
    if (attribute.featureName) {
      const featureName = attribute.featureName.toLowerCase();
      if (featureNameCount.has(featureName)) {
        featureNameCount.set(
          featureName,
          // @ts-ignore
          featureNameCount.get(featureName) + 1
        );
      } else {
        featureNameCount.set(featureName, 1);
      }
    }
  });

  let hasError = false;
  const featureErrors = featureList.map((attribute: FeatureAttributes) => {
    if (attribute.featureName) {
      // @ts-ignore
      if (featureNameCount.get(attribute.featureName.toLowerCase()) > 1) {
        hasError = true;
        return { featureName: 'Duplicate feature name' };
      } else {
        return undefined;
      }
    } else {
      hasError = true;
      // @ts-ignore
      return {
        featureName: 'Required',
      };
    }
  });
  return hasError ? { featureList: featureErrors } : undefined;
};

export const generateInitialFeatures = (detector: Detector): FeaturesFormikValues[] => {
  const featureUiMetaData = get(detector, 'uiMetadata.features', []);
  const features = get(detector, 'featureAttributes', []);
  // @ts-ignore
  return features.map((feature: FeatureAttributes) => {
    return {
      ...featureUiMetaData[feature.featureName],
      ...feature,
      aggregationQuery: JSON.stringify(feature['aggregationQuery'], null, 4),
      aggregationOf: get(
        featureUiMetaData,
        `${feature.featureName}.aggregationOf`
      )
        ? [
          {
            label: get(
              featureUiMetaData,
              `${feature.featureName}.aggregationOf`
            ),
          },
        ]
        : [],
      featureType: get(
        featureUiMetaData,
        `${feature.featureName}.featureType`
      )
        ? get(featureUiMetaData, `${feature.featureName}.featureType`)
        : FEATURE_TYPE.CUSTOM,
    };
  });
};