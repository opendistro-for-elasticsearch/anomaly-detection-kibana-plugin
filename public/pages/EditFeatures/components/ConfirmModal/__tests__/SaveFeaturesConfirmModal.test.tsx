/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from 'react';
import { render } from '@testing-library/react';
import { SaveFeaturesConfirmModal } from '../SaveFeaturesConfirmModal';
import { SAVE_FEATURE_OPTIONS } from '../../../../EditFeatures/utils/constants';


describe('<SaveFeaturesConfirmModal /> spec', () => {
  test('renders the component with ready to start job', () => {
    const { container } = render(
      <SaveFeaturesConfirmModal
        readyToStartAdJob={true}
        saveFeatureOption={SAVE_FEATURE_OPTIONS.START_AD_JOB}
        onClose={jest.fn()}
        onCancel={jest.fn()}
        onConfirm={jest.fn()}
        onOptionChange={jest.fn()}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
  
});
