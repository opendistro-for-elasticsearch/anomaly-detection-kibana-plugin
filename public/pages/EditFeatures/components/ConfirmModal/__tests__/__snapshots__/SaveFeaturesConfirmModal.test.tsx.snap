// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SaveFeaturesConfirmModal /> spec renders the component with ready to start job 1`] = `
<div
  data-focus-guard="true"
  style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
  tabindex="0"
/>
`;
